core/src/main/scala/org/apache/spark/sql/execution/row/RowDMLExec.scala

```diff
@@ -118,22 +121,39 @@ case class RowInsertExec(_child: SparkPlan, upsert: Boolean,
     val structFieldClass = classOf[StructField].getName
     val batchSize = connProps.executorConnProps
         .getProperty("batchsize", "1000").toInt
-    val numInsertedRowsMetric = if (onExecutor) null
-    else metricTerm(ctx, "numInsertedRows")
-    val numInsertions = ctx.freshName("numInsertions")
+    val numOpRowsMetric = if (onExecutor) null else metricTerm(ctx, s"num${opType}Rows")
+    val numOperations = ctx.freshName("numOperations")
     s"""
        |final $structFieldClass[] $schemaFields = $schemaTerm.fields();
        |${CodeGeneration.genStmtSetters(tableSchema.fields,
           connProps.dialect, input, stmt, schemaFields, ctx)}
        |$rowCount++;
        |$stmt.addBatch();
        |if (($rowCount % $batchSize) == 0) {
-       |  final int $numInsertions = $stmt.executeBatch().length;
-       |  $result += $numInsertions;
-       |  ${if (numInsertedRowsMetric eq null) ""
-            else s"$numInsertedRowsMetric.${metricAdd(numInsertions)};"}
+       |    final int $numOperations = $stmt.executeBatch().length;
+       |    $result += $numOperations;
+       |    ${if (numOpRowsMetric eq null) ""
+            else s"$numOpRowsMetric.${metricAdd(numOperations)};"}
        |  $rowCount = 0;
        |}
     """.stripMargin
   }
+
+  override def opType: String = if (upsert) {
+    "Upserted"
```

Comment:
Perhaps the name can be "Put" because it is different from a real upsert (arg can be renamed too).